File: algorithms-divide-conquer/week1/python/week1.py
import math


class Week1:

    @staticmethod
    def split(num):
        num_str = str(num)
        split_digit = 0
        if (len(num_str) % 2) == 0:
            '''Even'''
            split_digit = len(num_str) / 2
        else:
            '''Odd'''
            split_digit = len(num_str) / 2 + 1

        return [
            int( num_str[0:split_digit] ),
            int( num_str[split_digit:len(str(num_str))] )
        ]

    @staticmethod
    def karatsuba(int1, int2):
        if int1 < 10 or int2 < 10:
            return int1 * int2

        n = len(str(int1))
        n2 = len(str(int2))
        if n < n2:
            n = n2

        n2 = math.floor(n / 2);

        [a, b] = Week1.split(int1);
        [c, d] = Week1.split(int2);

        ac = Week1.karatsuba(a, c);
        bd = Week1.karatsuba(b, d);
        abcd = Week1.karatsuba(a + b, c + d);
        magic = abcd - ac - bd;

        return long(ac * 10 ** (2 * n2) + magic * 10 ** n2 + bd);

File: algorithms-divide-conquer/week1/week1.js
export default {
    split: function(num){
        const num_str = num.toString()
        var split_digit = 0
        if( ( num_str.length % 2 ) === 0 ){
            // Even
            split_digit = num_str.length / 2
        } else {
            // Odd
            split_digit = ( num_str.length / 2 ) + 1
        }
        return [
            parseInt( num_str.substring(0, split_digit) ),
            parseInt( num_str.substring(split_digit, split_digit.length) )
        ]
    },
    karatsuba: function(int1, int2){
        if( int1 < 10 || int2 < 10 ){
            return int1 * int2
        }

        const n = Math.max(int1.toString().length, int2.toString().length);
        const n2 = Math.floor(n / 2);

        const [a, b] = this.split(int1);
        const [c, d] = this.split(int2);

        const ac = this.karatsuba(a, c);
        const bd = this.karatsuba(b, d);
        const abcd = this.karatsuba(a + b, c + d);
        const magic = abcd - ac - bd;

        return ac * 10 ** (2 * n2) + magic * 10 ** n2 + bd;
    }
}